### DISCUSSION TITLE
Need help to integrate with traccar

### DISCUSSION THREAD

**Discussion by mikedolx** [owntracks/android#1665] [Q&A] [answered]
Hi,

TL;DR: I'm getting a http-error 400, when trying to send location updates to my traccar server.

i'd like to evaluate Owntracks with my traccar server. So far i have prepared my server to enable the traccar protocol. Furthermore, I have setup a reverse proxy to terminate the port on a specific domain like `https://mytraccar-server.com/owntracks`.

I have read the docs [here](https://owntracks.org/booklet/features/traccar/#traccar). I understand, that i somehow need to provide an identifier for traccar to identifiy my server. I still haven't fully understood, how to setup this identifier. I tried to add an extra property to my device like `tid=MI`.

I have entered this id under tracker-id in my owntracks app.

Now, when i try to send location updates i get an http-error 400, which according by the docs means the id probably does not match. But i don't know what i'm supposed to configure.

Can anyone give me a hint?

Thanks and BR;

**Comment by jpmens** [comment]
iirc, Android doesn't (yet) know about the "topic" variable in http mode, but you can configure a TID (tracker-id) which is a 2-character string. This string has to match the Traccar device name you configure there.

**Reply by mikedolx** [reply]
Cool, thanks!

But, the part i don't understand is, where to put the tid in the android client and the traccar server.

On my android client i have this options:
- device-id
- tracker-id

On the server side i have added the custom property `tid` all lower case. But there is also the device-id, which is used by the original traccar client.

**Reply by growse** [reply]
@jpmens have we got an open bug for topic in HTTP mode?

**Reply by mikedolx** [reply]
That's my exported owntracks configuration:

```json
{
  "_type" : "configuration",
  "waypoints" : [ ],
  "_build" : 420412000,
  "autostartOnBoot" : true,
  "cmd" : true,
  "connectionTimeoutSeconds" : 30,
  "debugLog" : true,
  "deviceId" : "MD",
  "dontReuseHttpClient" : false,
  "enableMapRotation" : true,
  "experimentalFeatures" : [ ],
  "fusedRegionDetection" : true,
  "ignoreInaccurateLocations" : 0,
  "ignoreStaleLocations" : 0.0,
  "locatorDisplacement" : 500,
  "locatorInterval" : 900,
  "locatorPriority" : 2,
  "mapLayerStyle" : "GoogleMapDefault",
  "mode" : 3,
  "monitoring" : 1,
  "moveModeLocatorInterval" : 10,
  "notificationEvents" : true,
  "notificationGeocoderErrors" : true,
  "notificationHigherPriority" : false,
  "notificationLocation" : true,
  "opencageApiKey" : "",
  "osmTileScaleFactor" : 1.0,
  "password" : "***",
  "pegLocatorFastestIntervalToInterval" : false,
  "ping" : 30,
  "pubExtendedData" : true,
  "remoteConfiguration" : false,
  "reverseGeocodeProvider" : "Device",
  "showRegionsOnMap" : false,
  "tid" : "MD",
  "url" : "http://192.168.2.233:5144",
  "username" : "***"
}
```

Meanwhile i have created a device on my traccar server with the following settings. No additional propeties.

![Screenshot_20240401-205119_Trebuchet.png](https://github.com/owntracks/android/assets/15738117/0c8175fa-37c5-44e8-9642-eec7611c7d58)


The tail of my android log is as follows:

```log
2024-04-01 22:16:19.326 D MessageProcessorEndpointHttp: url:http://192.168.2.233:5144/, messageId:1711997105425-d69bf2
2024-04-01 22:16:19.634 E MessageProcessorEndpointHttp: java.lang.Exception: HTTP request failed. Status: 400
	at org.owntracks.android.services.MessageProcessorEndpointHttp.sendMessage(SourceFile:528)
	at org.owntracks.android.services.MessageProcessor.lambda$sendAvailableMessages$0(Unknown Source:2)
	at org.owntracks.android.services.MessageProcessor.$r8$lambda$bVYaIhYhXtJCzBaSxWOGtcUu6f8(Unknown Source:0)
	at org.owntracks.android.services.MessageProcessor$$ExternalSyntheticLambda1.call(Unknown Source:4)
	at java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at android.os.Handler.handleCallback(Handler.java:942)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.os.HandlerThread.run(HandlerThread.java:67)

2024-04-01 22:16:19.637 D MessageProcessor: message:HTTP code 400, 
2024-04-01 22:16:19.639 E MessageProcessor: Message delivery failed. queueLength: 13, messageId: 1711997105425-d69bf2
2024-04-01 22:16:19.640 W MessageProcessor: Error sending message. Re-queueing
2024-04-01 22:16:19.649 I MessageProcessor: Waiting for 120 s before retrying
```

**Reply by jpmens** [reply]
what does Traccar logging say?

**Reply by mikedolx** [reply]
I caught this one with the traccar id as the device-id

```
2024-04-03 04:02:02 WARN: Unknown device - owntracks/mytraccarusername/358842 (192.168.2.204)
```

**Reply by mikedolx** [reply]
Second attempt with the tid

```
2024-04-03 04:04:28  WARN: Unknown device - owntracks/Michael/MD (192.168.127.204)
```

**Reply by ckrey** [reply]
put in exactly this 
> owntracks/Michael/MD

in traccar

**Reply by mikedolx** [reply, marked as answer]
That seems to work. At least the owntracks client is getting a http code 200. But apperently the response cannot be parsed by the client:

```
2024-04-03 08:19:57.444 I MessageProcessorEndpointHttp: Message id=1712124589032-f7e554 sent in 144ms
2024-04-03 08:19:57.445 E MessageProcessorEndpointHttp: JsonParseException HTTP status: 200
```
The log on the server side looks good:

```
2024-04-03 08:29:02 INFO: Event id: owntracks/Michael/MD, time: 2024-04-03 08:29:01, type: deviceOnline, notifications: 1
```

**Reply by ckrey** [reply]
you may ignore this. If anything is replied by the server, the app checks if it is valid JSON and proceeds if this is the case.
From traccar, nothing is returned.

**Reply by mikedolx** [reply]
Hey, 

it's me again. I have lately tried again to usw owntrack, but again i'm getting the http 400 error:

```
2024-04-23 06:01:34.287 I TimberInMemoryLogTree: Logs cleared
2024-04-23 06:01:45.503 E MessageProcessorEndpointHttp: java.lang.Exception: HTTP request failed. Status: 400
	at org.owntracks.android.services.MessageProcessorEndpointHttp.sendMessage(SourceFile:528)
	at org.owntracks.android.services.MessageProcessor.lambda$sendAvailableMessages$0(Unknown Source:2)
	at org.owntracks.android.services.MessageProcessor.$r8$lambda$bVYaIhYhXtJCzBaSxWOGtcUu6f8(Unknown Source:0)
	at org.owntracks.android.services.MessageProcessor$$ExternalSyntheticLambda1.call(Unknown Source:4)
	at java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at android.os.Handler.handleCallback(Handler.java:942)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.os.HandlerThread.run(HandlerThread.java:67)

2024-04-23 06:01:45.506 E MessageProcessor: Message delivery failed. queueLength: 2429, messageId: 1713818774746-aa75c4
2024-04-23 06:01:45.507 W MessageProcessor: Error sending message. Re-queueing
2024-04-23 06:01:45.519 I MessageProcessor: Waiting for 120 s before retrying
```

I set up everything as descried above. I have added the I'd `owntracks/MD/Michael` , but the server still complains about the ID being not recognized.

**Reply by jpmens** [reply]
what did you change meanwhile? Three weeks ago you reported Traccar saying the device is online... Did you upgrade some software, and if so what to which?

**Reply by mikedolx** [reply]
AS it seems this is an issue with my reverse proxy. I remember,  the last successful attempt was done using VPN and the IP/Port as hostname in own tracks. Now, I tried it with my domain and the reverse proxy.

I'm using s reverse proxy for my traccar server. For the own tracks protocol I use a route, so don't have to open another port in my router and everything can be passed through the SSL/443 port. 

Apparently, the reverse proxy config was wrong. It seems I forgot the trailing slash `/` for my owntracks-path. It seems to work for now. I'm going to do a test drive and report back. Eventually I'll post the reverse proxy snippet.

**Reply by mikedolx** [reply]
Jepp it works! 

Here's how i set uo my Route in the Apache reverse proxy

```
ProxyPass /owntracks/ http://192.168.2.200:5144/
ProxyPassReverse /owntracks/ http://192.168.2.200:5144/
```

Cheers!

**Comment by jpmens** [comment]
on Android TID is the tracker-id.
on Traccar, you put your TID in the device ID field of the Traccar device configuration.

**Comment by jpmens** [comment]
there was https://github.com/owntracks/android/issues/760 and https://github.com/owntracks/android/issues/512

At the time "we" decided not to follow up on this, and iirc this is the first time in over 5 years that someone wants to connect our Android to Traccar ... ;)

**Reply by jpmens** [reply]
@growse I see your question via mail but not here …

**Reply by mikedolx** [reply]
@jpmens is there a way to use an MQTT-Broker in between, and then transform the payload to push it into the traccar server?

Edit: Maybe with Nodered as the middleware in between?

**Reply by jpmens** [reply]
sure that's possible, but it would mean more moving parts ...
